owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -0,0 +1,199 @@
+    this.notStartedPieces = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    this.totalPieces = torrent.getPieces().size();
+
+    for (int idx = 0; idx < totalPieces; idx++) {
+      this.notStartedPieces.add(idx);
+    }
+
+    this.peers = new ConcurrentHashMap<>();
+
+    // initPeers();
+  }
+
+  private void initPeers(List<Peer> peerList) {
+    for (Peer peer : peerList) {
+      try {
+        PeerConnector connector =
+            SocketConnector.makeInitialConnection(peer, this, networkStorageAdapter);
+        addPeer(connector, peer);
+        // TODO: revise
+        connector.writeMessage(new PayloadlessMessage(PeerMessage.MessageType.INTERESTED));
+      } catch (IOException e) {
+        log.error(String.format("Error connecting peer id=%s", peer.getPeerID().toString()));
+      }
+    }
+  }
+
+  public void startDownloadingAsynchronously() {
+    // TODO: make thread pool scheduler
+    new Thread(this).start();
+  }
+
+  public float getProgressPercent() {
+    return completedPieces.size() * 1.0f / totalPieces;
+  }
+
+  // For handshake listener
+  // Later if get updated peerlist from tracker.
+  public void addPeer(PeerConnector connector, Peer peer) {
+    this.peers.put(peer, connector);
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (var pair : peers.entrySet()) {
+      pair.getValue().close();
+    }
+  }
+
+  private void requestPieceFromPeer(Peer peer, int pieceNum, int blockNum, int blockSize) {
+    PeerConnector peerConnector = peers.get(peer);
+    if (peerConnector == null) {
+      return;
+    }
+
+    // TODO: fix int casting.
+    try {
+      peerConnector.writeMessage(
+          PieceActionMessage.makeRequestMessage(pieceNum, blockNum * blockSize, blockSize));
+    } catch (IOException e) {
+      log.error(e);
+    }
+  }
+
+  @Override
+  public void run() {
+    while (!uncompletedPieces.isEmpty()) {
+      // Request a missing piece from each Peer
+      List<Peer> connections = new ArrayList<>(peers.keySet());
+      Collections.shuffle(connections);
+
+      for (PieceStatus progress : uncompletedPieces.values()) {
+        for (int i = 0; i < progress.status.size(); i++) {
+          if (connections.isEmpty()) break;
+
+          if (progress.status.get(i).get() != BLOCK_NOT_STARTED) {
+            continue;
+          }
+
+          requestPieceFromPeer(connections.remove(0), progress.pieceIndex, i, progress.blockLength);
+        }
+        if (connections.isEmpty()) break;
+      }
+      // Thread.sleep(1000);
+    }
+  }
+
+  /**
+   * Reports the progress of a block
+   *
+   * @param blockInfo block info
+   * @return false if another thread already started to download that block or the block size is not
+   *     expected
+   */
+  public boolean reportAndValidatePieceInProgress(FileBlockInfo blockInfo) {
+    // TODO: fix int casting, and also add handling when piece length isn't a multiple of default
+    //  block num
+    PieceStatus status =
```

Comment:
We should just initialzie these at the beginning, so the hashmap should always have them and no need to generate here

Reply:
This was the optimization we talked about, where we instantiate the block status array only after the block is touched. (It turned out easier to implement than I expected lol)

Reply:
Oh oops I forgot to remove the index from not started...

Reply:
ah fair enough